### PR TITLE
RPL (RFC 6550, RFC 6551) support

### PR DESCRIPTION
**Pull request by nyrahul**
**Checklist:**

-   [x] I have checked [CONTRIBUTING.md](https://github.com/secdev/scapy/blob/master/CONTRIBUTING.md) (esp. section submitting-pull-requests)
-   [x] I squashed commits belonging together
-   [x] I added unit tests 
-   [x] I executed the regression tests (using `cd test && ./run_tests`)
-   [x] The PR is complete.

## Description
Support for RPL routing protocol used in Low power and lossy networks (RFC 6550/RFC 6551). Also supports [draft-ietf-roll-efficient-npdao](https://datatracker.ietf.org/doc/draft-ietf-roll-efficient-npdao/).

## Examples
1. RPL DIS message with Solicited Information Option
```
msg=IPv6() / ICMPv6RPL() / RPLDIS() / RPLOptSolInfo()
```
2. RPL DAO message with multiple target, multiple transit options
```
msg=IPv6(src="fe80::1", dst="fe80::2") / ICMPv6RPL() / RPLDAO()/\
        RPLOptTgt(plen=128,prefix="aaaa::1") / RPLOptTIO(pathseq=11)/\
        RPLOptTgt(plen=128,prefix="bbbb::1") / RPLOptTIO(pathseq=22)/\
        RPLOptTgt(plen=128,prefix="cccc::1") / RPLOptTIO(pathseq=33)
```
3. RPL DIO message with DAG Metric Container containing Link ETX metric
```
msg=ICMPv6RPL() / RPLDIO() / RPLOptDAGMC() / RPLLinkETX()
```

All the produced messages are tested/dissected fine with Wireshark.

### CODE REVIEW

File: scapy/contrib/rpl.py [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+
+
+rplopts = {
+}
+
+
+class RPLGuessOption:
+    name = "Dummy RPL Option class that implements guess_payload_class()"
+
+    def guess_payload_class(self, p):
```
Comment (gpotter2) [inline scapy/contrib/rpl.py:59]: Instead of using `guess_payload_class`, we prefer to use multiple `bind_layers` calls.

You can keep the dummy and do all your `bind_layers` call on it.

File: scapy/contrib/rpl.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    name = "Destination Cleanup Object Acknowledgement"
+    fields_desc = [XShortField("cksum", None),
+                   # Base Object
+                   ByteField("RPLInstanceID", 50),
+                   BitField("D", 0, 1),
+                   BitField("flags", 0, 7),
+                   ByteField("dcoseq", 1),
+                   ByteField("status", 0),
+                   _OptDODAGIDField("dodagid", None)]
+    overload_fields = {RPL: {"code": 8}}
```
Comment (gpotter2) [inline scapy/contrib/rpl.py:269]: This can be replaced by a `bind_top_down` call (it's cleaner). I know the user of overload_fields is common in Scapy but it's not something we want to keep.

File: scapy/layers/inet6.py [hunk 3]
```diff
@@ -2600,6 +2604,43 @@ def _niquery_guesser(p):
...
+icmp6rplcodes = {
+}  # filled in contrib/rpl.py
+
+
+class _RPLGuessPayload:
+    name = "Dummy RPL class that implements guess_payload_class()"
+
+    def guess_payload_class(self, p):
+        if len(p) > 1:
+            return icmp6rplcodes.get(self.code, Raw)
```
Comment (gpotter2) [inline scapy/layers/inet6.py:2622]: Same

File: scapy/layers/inet6.py [hunk 4]
```diff
@@ -3966,6 +4007,7 @@ def _load_dict(d):
         d[k] = _get_cls(v)
 
 
+_load_dict(icmp6rplcodes)
```
Comment (gpotter2) [inline scapy/layers/inet6.py:4010]: This will never be called: the contrib module will always be called after `inet6`, meaning this will always be empty.

File: scapy/contrib/rpl.py (hunk 1 above)
Comment (nyrahul) [reply scapy/contrib/rpl.py]: Fixed: `bind_layers` is actually better to use. More readable.

File: scapy/contrib/rpl.py (hunk 2 above)
Comment (nyrahul) [reply scapy/contrib/rpl.py]: Fixed: Removed `overload_fields` from all over RPL.

File: scapy/layers/inet6.py (hunk 3 above)
Comment (nyrahul) [reply scapy/layers/inet6.py]: Fixed.

File: scapy/layers/inet6.py (hunk 4 above)
Comment (nyrahul) [reply scapy/layers/inet6.py]: Fixed: removed icmp6rplcodes and some other relevant dicts.

File: scapy/contrib/rpl.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+              5: "RPL Target",
+              6: "Transit Information",
+              7: "Solicited Information",
+              8: "Prefix Information Option",
+              9: "Target Descriptor",
+              10: "P2P Route Discovery"}
+
+
+class _RPLGuessOption(Packet):
+    name = "Dummy RPL Option class that implements guess_payload_class()"
```
Comment (gpotter2) [inline scapy/contrib/rpl.py:57]: Minor detail but this docstrings is now wrong

File: scapy/contrib/rpl.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    Control Message: Destination Advertisement Object (DAO)
+    """
+    name = "Destination Advertisement Object"
+    fields_desc = [ByteField("RPLInstanceID", 50),
+                   BitField("K", 0, 1),
+                   BitField("D", 0, 1),
+                   BitField("flags", 0, 6),
+                   ByteField("reserved", 0),
+                   ByteField("daoseq", 1),
+                   _OptDODAGIDField("dodagid", None)]
```
Comment (gpotter2) [inline scapy/contrib/rpl.py:247]: Can't you use a `ConditionalField` here instead ? The warning isn't that useful

File: scapy/contrib/rpl.py [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+    fields_desc = [ByteEnumField("otype", 0x01, RPLOPTSSTR),
+                   FieldLenField("optlen", None, length_of="optdata", fmt="B"),
+                   StrLenField("optdata", "",
+                               length_from=lambda pkt: pkt.optlen)]
+
+
+# RPL Control Message Handling
+
+
+class DIS(_RPLGuessOption, Packet):
```
Comment (gpotter2) [inline scapy/contrib/rpl.py:192]: The Packet is useless. Same below

File: scapy/contrib/rpl.py [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+bind_layers(RPL, DIS, code=0)
+bind_layers(RPL, DIO, code=1)
+bind_layers(RPL, DAO, code=2)
+bind_layers(RPL, DAOACK, code=3)
+bind_layers(RPL, DCO, code=7)
+bind_layers(RPL, DCOACK, code=8)
+
+
+# https://www.iana.org/assignments/rpl/rpl.xhtml#control-message-options
+for msg in [DIS, DIO, DAO, DAOACK, DCO, DCOACK]:
```
Comment (gpotter2) [inline scapy/contrib/rpl.py:302]: You can use `bind_layers(_RPLGuessOption, ...)` instead of doing it for each since they all inherit from it

File: scapy/layers/inet6.py [hunk 9]
```diff
@@ -3995,3 +4024,4 @@ def _load_dict(d):
 bind_layers(IPv6, IPv6, nh=socket.IPPROTO_IPV6)
 bind_layers(IPv6, IP, nh=socket.IPPROTO_IPIP)
 bind_layers(IPv6, GRE, nh=socket.IPPROTO_GRE)
+bind_layers(IPv6, RPL, {"nh": 58, "dst": "ff02::1a"})
```
Comment (gpotter2) [inline scapy/layers/inet6.py:4027]: You can use keywords, it's cleaner: `nh=58, dst=...`

File: scapy/contrib/rpl_metrics.py [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+RFC 6551 - Routing Metrics Used for Path Calculation in LLNs
+"""
+
+import struct
+from scapy.compat import orb
+from scapy.packet import Packet, bind_layers
+from scapy.fields import ByteEnumField, ByteField, ShortField, BitField, \
+    BitEnumField, FieldLenField, StrLenField, IntField
+from scapy.layers.inet6 import _PhantomAutoPadField, _OptionsField
+from scapy.contrib.rpl import rploptsstr, DIS, DIO, DAO, DAOACK, DCO, DCOACK
```
Comment (gpotter2) [inline scapy/contrib/rpl_metrics.py:33]: `rploptsstr` isn't defined. This import fails

Comment (nyrahul) [reply scapy/contrib/rpl_metrics.py]: Sorry, mistake. Committed in haste. Fixed.

File: scapy/layers/inet6.py (hunk 9 above)
Comment (nyrahul) [reply scapy/layers/inet6.py]: Cleaned up.

File: scapy/contrib/rpl.py (hunk 8 above)
Comment (nyrahul) [reply scapy/contrib/rpl.py]: Actually there are two things:
1. RPL Control Messages (DIS/DIO/DAO/DAO...)
2. RPL Control Options (Pad1/N, Target, TIO, PIO, RIO ...)

The messages can happen one or more options. So I need DIS/DIO.. to be represented by a different dummy class. I am now using _RPLGuessMsgType to represent Messages ... and the previously used _RPLGuessOption for options.
Kindly check if this makes sense.

File: scapy/contrib/rpl.py (hunk 7 above)
Comment (nyrahul) [reply scapy/contrib/rpl.py]: Removed Packet.

File: scapy/contrib/rpl.py (hunk 6 above)
Comment (nyrahul) [reply scapy/contrib/rpl.py]: Yep, `ConditionalField` makes perfect sense.  Replaced.

File: scapy/contrib/rpl.py (hunk 5 above)
Comment (nyrahul) [reply scapy/contrib/rpl.py]: Fixed the docstring. Thanks

**Comment by gpotter2** [comment]
Looks good to me ! Now you just have to make the tests pass.

File: scapy/contrib/rpl_metrics.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+    Dummy unknown metric/constraint
+    """
+    name = 'Unknown DAGMC Object Option'
+    fields_desc = [ByteEnumField("otype", 3, DAGMC_OBJTYPE),
+                   FieldLenField("olen", None, length_of="odata", fmt="B"),
+                   StrLenField("odata", "",
+                               length_from=lambda pkt: pkt.olen)]
+
+    @classmethod
+    def dispatch_hook(cls, _pkt=None, *_, **kargs):
```
Comment (guedou) [inline scapy/contrib/rpl_metrics.py:77]: It seems that this is never reached by your tests (https://codecov.io/gh/secdev/scapy/pull/2663/diff?src=pr&el=tree#diff-c2NhcHkvY29udHJpYi9ycGxfbWV0cmljcy5weQ==). Could you find a way to test this?

File: scapy/contrib/rpl.py [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+
+class _RPLGuessMsgType(Packet):
+    name = "Dummy RPL Message class"
+
+
+class _RPLGuessOption(Packet):
+    name = "Dummy RPL Option class"
+
+
+class OptRIO(_RPLGuessOption):
```
[the remark this answers is not in the document]
Comment (nyrahul) [reply scapy/contrib/rpl.py:78]: Thanks @guedou for the review and feedback.

Regarding renaming, I would like to put up an explanation before I handle the change. I think renaming RPL to ICMPv6RPL() makes sense but doing it for all the RPL messages and options may not be right. RPL options {OptDAGMC, Opt*} are not tied to ICMPv6 at all so renaming those does not make sense.

Technically it is possible to use RPL above the transport layer like UDP (just like RIP[RFC2453]) in the future. This is not true for the ICMPv6 messages you referred to.

From a code compliance standpoint, the RPL/ICMPv6RPL is defined in `inet6.py`. But all the RPL messages and options are segregated in their own code file (`rpl.py` and `rpl_metrics.py`) and this was done to ensure that these messages/options could be used on top of other underlayers.

Readability is another concern... Consider the example I used to raise this PR:

Currently:
```
msg=IPv6()/RPL()/DIO()/OptDAGMC()/LinkETX()
```

prefixing ICMPv6RPL 
```
msg=IPv6()/ICMPv6RPL()/ICMPv6RPLDIO()/ICMPv6RPLOptDAGMC()/ICMPv6RPLLinkETX()
```

To sum up: I believe, only RPL should be renamed to ICMPv6RPL().
```
msg=IPv6()/ICMPv6RPL()/DIO()/OptDAGMC()/LinkETX()
```

I will update based on your response. PLMK, Thanks.

File: scapy/contrib/rpl.py [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+    """
+    Control Option: RPL Target Descriptor
+    """
+    name = "RPL Target Descriptor"
+    fields_desc = [ByteEnumField("otype", 9, RPLOPTSSTR),
+                   ByteField("len", 4),
+                   IntField("descriptor", 0)]
+
+
+class Pad1(_RPLGuessOption):
```
[the remark this answers is not in the document]
Comment (nyrahul) [reply scapy/contrib/rpl.py:188]: Pad1 and PadN are defined as `_OTypeField()` in `inet6.py` and are hbh option specific. I initially used those but they ended up showing hbh specific representation. And hence I changed.

File: scapy/layers/inet6.py [hunk 14]
```diff
@@ -3995,3 +4024,4 @@ def _load_dict(d):
 bind_layers(IPv6, IPv6, nh=socket.IPPROTO_IPV6)
 bind_layers(IPv6, IP, nh=socket.IPPROTO_IPIP)
 bind_layers(IPv6, GRE, nh=socket.IPPROTO_GRE)
+bind_layers(IPv6, RPL, nh=58, dst="ff02::1a")
```
[the remark this answers is not in the document]
Comment (nyrahul) [reply scapy/layers/inet6.py:4027]: `ICMPv6MLQuery()` seems to be using `overload_fields` for overriding the dst-address and next-header. Actually I used the same initially but was suggested to use `bind_layers` by @gpotter2

File: scapy/contrib/rpl.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (guedou) [reply scapy/contrib/rpl.py:78]: I will prefer to have:
`msg = IPv6() / ICMPv6RPL() / RPLDIO() / RPLOptDAGMC() / RPLLinkETX()`

That way, completion could easily be used to display the RPL related packets.

File: scapy/contrib/rpl.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (guedou) [reply scapy/contrib/rpl.py:188]: I understand. Thanks.

File: scapy/contrib/rpl.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (nyrahul) [reply scapy/contrib/rpl.py:78]: Sounds good. Thanks.

File: scapy/layers/inet6.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/layers/inet6.py:4027]: I meant `bind_top_down`.

IPv6 uses a lot of  `overload_fields` (which could be replaced by `bind_top_down` calls). It has the exact same effect though i prefer the second option because it "looks" fancier. But in this case you might want to remain consistent as guedou said.

[the remark this answers is not in the document]
Comment (nyrahul) [reply scapy/layers/inet6.py]: Thanks @gpotter2

File: scapy/layers/inet6.py [hunk 15]
```diff
@@ -2600,6 +2604,31 @@ def _niquery_guesser(p):
...
+            2: "DAO",
+            3: "DAO-ACK",
+            # 4: "P2P-DRO",
+            # 5: "P2P-DRO-ACK",
+            # 6: "Measurement",
+            7: "DCO",
+            8: "DCO-ACK"}
+
+
+class RPL(_ICMPv6):   # RFC 6550
```
[the remark this answers is not in the document]
Comment (nyrahul) [reply scapy/layers/inet6.py]: Fixed

File: scapy/contrib/rpl.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (nyrahul) [reply scapy/contrib/rpl.py]: Fixed

File: scapy/contrib/rpl.py [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+    fields_desc = [ByteEnumField("otype", 0x01, RPLOPTSSTR),
+                   FieldLenField("optlen", None, length_of="optdata", fmt="B"),
+                   StrLenField("optdata", "",
+                               length_from=lambda pkt: pkt.optlen)]
+
+
+# RPL Control Message Handling
+
+
+class DIS(_RPLGuessMsgType, _RPLGuessOption):
```
[the remark this answers is not in the document]
Comment (nyrahul) [reply scapy/contrib/rpl.py]: Fixed as per the agreed naming convention in the comment above.

File: test/contrib/rpl.uts [hunk 17]
```diff
@@ -0,0 +1,70 @@
+% RPL layer test campaign
+
++ Syntax check
+= Import the RPL layer
+from scapy.contrib.rpl import *
```
[the remark this answers is not in the document]
Comment (nyrahul) [reply test/contrib/rpl.uts]: using load_contrib for both rpl/rpl_metrics.

File: test/contrib/rpl.uts [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+from scapy.contrib.rpl_metrics import *
```
[the remark this answers is not in the document]
Comment (nyrahul) [reply test/contrib/rpl.uts]: fixed

File: scapy/contrib/rpl_metrics.py (hunk 11 above)
Comment (nyrahul) [reply scapy/contrib/rpl_metrics.py:77]: Vastly improved UT in the update.

File: scapy/layers/inet6.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (nyrahul) [reply scapy/layers/inet6.py]: Hi @gpotter2 , Can you please review the change in rpl.py again? I could not get the dissection to work without `guess_payload_class`.

Dissection works perfectly fine with `guess_payload_class`. I realized this (that dissection does not work with bind_layer but works with guess_payload_class) after adding more Unit tests.

[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/layers/inet6.py]: It's fine :+1:

File: scapy/layers/inet6.py [hunk 19]
```diff
@@ -1202,6 +1202,8 @@ def fragment6(pkt, fragSize):
                  151: "ICMPv6MRD_Advertisement",
                  152: "ICMPv6MRD_Solicitation",
                  153: "ICMPv6MRD_Termination",
+                 # 154: Do Me - FMIPv6 Messages - RFC 5568
```
Comment (guedou) [inline scapy/layers/inet6.py:1205]: Is this comment needed? I doubt that FMIPv6 will ever be implemented =)

**Comment by guedou** [comment]
Awesome work! Thanks for your patience and your fixes.
